crates/rattler_installs_packages/src/sdist.rs

```diff
@@ -0,0 +1,118 @@
+use crate::core_metadata::WheelCoreMetadata;
+use crate::utils::ReadAndSeek;
+use crate::{Artifact, MetadataArtifact, SDistFormat, SDistName};
+use flate2::read::GzDecoder;
+use miette::{miette, IntoDiagnostic};
+use parking_lot::Mutex;
+use std::ffi::OsStr;
+use std::io::Read;
+use std::path::Path;
+use std::str::FromStr;
+use tar::Archive;
+
+/// Represents a source distribution artifact.
+pub struct SDist {
+    /// Name of the source distribution
+    name: SDistName,
+
+    /// Source dist archive
+    archive: Mutex<Archive<Box<dyn Read + Send>>>,
+}
+
+impl SDist {
+    /// Create this struct from a path
+    #[allow(dead_code)]
+    pub fn from_path(path: &Path) -> miette::Result<Self> {
+        let file_name = path
+            .file_name()
+            .and_then(OsStr::to_str)
+            .ok_or_else(|| miette::miette!("path does not contain a filename"))?;
+        let name = SDistName::from_str(file_name).into_diagnostic()?;
+        let bytes = std::fs::File::open(path).into_diagnostic()?;
+        Self::new(name, Box::new(bytes))
+    }
+
+    /// Read .PKG-INFO from the archive
+    pub fn read_package_info(&self) -> miette::Result<(Vec<u8>, WheelCoreMetadata)> {
+        let mut archive = self.archive.lock();
+
+        // Loop over entries
+        for entry in archive.entries().into_diagnostic()? {
+            let mut entry = entry.into_diagnostic()?;
+
+            // Find PKG-INFO and parse this file
+            if entry.path().into_diagnostic()?.ends_with("PKG-INFO") {
+                let mut bytes = Vec::new();
+                entry.read_to_end(&mut bytes).into_diagnostic()?;
+                let metadata = Self::parse_metadata(&bytes)?;
+                return Ok((bytes, metadata));
```

Comment:
This should only use this metadata from the PKG-INFO file if the metadata version is 2.2 or greater. Using on older versions means that you'd rely on the metadata not required to be reliable.